source/ada/lsp-ada_handlers.adb

```diff
@@ -2647,81 +2648,96 @@ package body LSP.Ada_Handlers is
                    Get_Open_Document (Self, Value.textDocument.uri);
 
+      Refs      : Laltools.Refactor.Rename.Renamable_References;
+
       procedure Process_Context (C : Context_Access);
       --  Process the rename request for the given context, and add
-      --  the results to response.
+      --  the results to response. If a rename problem is found, return a
+      --  response with an error.
 
       ---------------------
       -- Process_Context --
       ---------------------
 
       procedure Process_Context (C : Context_Access) is
-         Position : constant LSP.Messages.TextDocumentPositionParams :=
-                      (Value.textDocument, Value.position);
+         Position   : constant LSP.Messages.TextDocumentPositionParams :=
+           (Value.textDocument, Value.position);
 
-         Name_Node  : constant Name := Laltools.Common.Get_Node_As_Name
+         Node      : Ada_Node := C.Get_Node_At (Document, Position);
+         Name_Node : Name := Laltools.Common.Get_Node_As_Name
            (C.Get_Node_At (Document, Position));
 
-         Definition : Defining_Name;
-         Imprecise  : Boolean;
-         Empty      : LSP.Messages.TextEdit_Vector;
-         Count      : Cancel_Countdown := 0;
+         Empty     : LSP.Messages.TextEdit_Vector;
+
+         --  Analysis unit of the reference that is being added to Response.
+         Unit      : Analysis_Unit;
 
          procedure Process_Comments
            (Node : Ada_Node;
             Uri  : LSP.Messages.DocumentUri);
          --  Iterate over all comments and include them in the response when
-         --  they contain a renamed word
+         --  they contain a renamed word.
 
-         procedure Callback
-           (Node   : Libadalang.Analysis.Base_Id;
-            Kind   : Libadalang.Common.Ref_Result_Kind;
-            Cancel : in out Boolean);
+         procedure Process_Reference
+           (Unit_Filename : String;
+            Sloc_Range    : Langkit_Support.Slocs.Source_Location_Range);
+         --  Add a reference to Response if it has not been added yet.
 
-         --------------
-         -- Callback --
-         --------------
+         -----------------------
+         -- Process_Reference --
+         -----------------------
 
-         procedure Callback
-           (Node   : Libadalang.Analysis.Base_Id;
-            Kind   : Libadalang.Common.Ref_Result_Kind;
-            Cancel : in out Boolean)
+         procedure Process_Reference
+           (Unit_Filename : String;
+            Sloc_Range    : Langkit_Support.Slocs.Source_Location_Range)
          is
+            use type VSS.Unicode.UTF16_Code_Unit_Count;
+
             Location : constant LSP.Messages.Location :=
-              Get_Node_Location (Node.As_Ada_Node);
+              (uri     => LSP.Ada_Contexts.File_To_URI
+                 (To_Unbounded_Wide_String
+                      (To_Wide_String (Unit_Filename))),
+               span    => LSP.Messages.Span'
```

Comment:
how about introducing an utility function to transform a Langkit Sloc_Range to a LSP span?